drivers/oled/oled_driver.c

```diff
@@ -132,24 +165,116 @@ uint32_t oled_scroll_timeout;
 uint16_t oled_update_timeout;
 #endif
 
-// Internal variables to reduce math instructions
+#if defined(OLED_TRANSPORT_SPI)
+#    ifndef OLED_DC_PIN
+#        error "The OLED driver in SPI needs a D/C pin defined"
+#    endif
+#    ifndef OLED_CS_PIN
+#        error "The OLED driver in SPI needs a CS pin defined"
+#    endif
+#    ifndef OLED_SPI_MODE
+#        define OLED_SPI_MODE 3
+#    endif
+#    ifndef OLED_SPI_DIVISOR
+#        define OLED_SPI_DIVISOR 2
+#    endif
+#elif defined(OLED_TRANSPORT_I2C)
+#    if !defined(OLED_DISPLAY_ADDRESS)
+#        define OLED_DISPLAY_ADDRESS 0x3C
+#    endif
+#endif
+
+// Transmit/Write Funcs.
+__attribute__((weak)) bool oled_send_cmd(const uint8_t *data, uint16_t size) {
+#if defined(OLED_TRANSPORT_SPI)
+    if (!spi_start(OLED_CS_PIN, false, OLED_SPI_MODE, OLED_SPI_DIVISOR)) {
+        return false;
+    }
+    // Command Mode
+    writePinLow(OLED_DC_PIN);
+    // Send the commands
+    if (spi_transmit(&data[1], size - 1) != SPI_STATUS_SUCCESS) {
+        spi_stop();
+        return false;
+    }
+    spi_stop();
+    return true;
+#elif defined(OLED_TRANSPORT_I2C)
+    i2c_status_t status = i2c_transmit((OLED_DISPLAY_ADDRESS << 1), data, size, OLED_I2C_TIMEOUT);
+
+    return (status == I2C_STATUS_SUCCESS);
+#endif
+}
 
+__attribute__((weak)) bool oled_send_cmd_P(const uint8_t *data, uint16_t size) {
 #if defined(__AVR__)
-// identical to i2c_transmit, but for PROGMEM since all initialization is in PROGMEM arrays currently
-// probably should move this into i2c_master...
-static i2c_status_t i2c_transmit_P(uint8_t address, const uint8_t *data, uint16_t length, uint16_t timeout) {
-    i2c_status_t status = i2c_start(address | I2C_WRITE, timeout);
+#    if defined(OLED_TRANSPORT_SPI)
+    if (!spi_start(OLED_CS_PIN, false, OLED_SPI_MODE, OLED_SPI_DIVISOR)) {
+        return false;
+    }
+    spi_status_t status = SPI_STATUS_SUCCESS;
+    // Command Mode
+    writePinLow(OLED_DC_PIN);
+    // Send the commands
+    for (uint16_t i = 1; i < size && status >= 0; i++) {
+        status = spi_write(pgm_read_byte((const char *)&data[i]));
+    }
+    spi_stop();
+    return (status >= 0);
+#    elif defined(OLED_TRANSPORT_I2C)
+    i2c_status_t status = i2c_start((OLED_DISPLAY_ADDRESS << 1) | I2C_WRITE, OLED_I2C_TIMEOUT);
 
-    for (uint16_t i = 0; i < length && status >= 0; i++) {
-        status = i2c_write(pgm_read_byte((const char *)data++), timeout);
-        if (status) break;
+    for (uint16_t i = 0; i < size && status >= 0; i++) {
+        status = i2c_write(pgm_read_byte((const char *)data++), OLED_I2C_TIMEOUT);
     }
 
     i2c_stop();
 
-    return status;
+    return (status == I2C_STATUS_SUCCESS);
+#    endif
+#else
+    return oled_send_cmd(data, size);
+#endif
 }
+
+__attribute__((weak)) bool oled_send_data(const uint8_t *data, uint16_t size) {
+#if defined(OLED_TRANSPORT_SPI)
+    spi_start(OLED_CS_PIN, false, OLED_SPI_MODE, OLED_SPI_DIVISOR);
```

Comment:
As we have this error check in other places…
```suggestion
    if (!spi_start(OLED_CS_PIN, false, OLED_SPI_MODE, OLED_SPI_DIVISOR)) {
        return false;
    }
```